onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,562 @@
+  };
+
+  std::unique_ptr<Provider_IExecutionProvider_Router> Create_IExecutionProvider_Router(Provider_IExecutionProvider* outer, const std::string& type) override {
+    return onnxruntime::make_unique<Provider_IExecutionProvider_Router_Impl>(outer, type);
+  };
+
+  logging::Logger* LoggingManager_GetDefaultLogger() override {
+    return const_cast<logging::Logger*>(&logging::LoggingManager::DefaultLogger());
+  }
+
+  void* HeapAllocate(size_t size) override { return new uint8_t[size]; }
+  void HeapFree(void* p) override { delete reinterpret_cast<uint8_t*>(p); }
+
+  bool CPU_HasAVX2() override {
+    return CPUIDInfo::GetCPUIDInfo().HasAVX2();
+  }
+
+  bool CPU_HasAVX512f() override {
+    return CPUIDInfo::GetCPUIDInfo().HasAVX512f();
+  }
+
+  void LogRuntimeError(uint32_t session_id, const common::Status& status, const char* file, const char* function, uint32_t line) override {
+    return ::onnxruntime::LogRuntimeError(session_id, status, file, function, line);
+  }
+
+} provider_host_;
+
+struct ProviderLibrary {
+  ProviderLibrary(const char* filename) {
+    Env::Default().LoadDynamicLibrary(filename, &handle_);
+    if (!handle_)
+      return;
+
+    Provider* (*PGetProvider)();
+    Env::Default().GetSymbolFromLibrary(handle_, "GetProvider", (void**)&PGetProvider);
+
+    provider_ = PGetProvider();
+    provider_->SetProviderHost(provider_host_);
+  }
+
+  ~ProviderLibrary() {
+    Env::Default().UnloadDynamicLibrary(handle_);
+  }
+
+  Provider* provider_{};
+  void* handle_{};
+};
+
+// This class translates the IExecutionProviderFactory interface to work with the interface providers implement
+struct IExecutionProviderFactory_Translator : IExecutionProviderFactory {
+  IExecutionProviderFactory_Translator(std::shared_ptr<Provider_IExecutionProviderFactory> p) : p_{p} {}
+
+  std::unique_ptr<IExecutionProvider> CreateProvider() override {
+    auto provider = p_->CreateProvider();
+    return std::unique_ptr<IExecutionProvider>(static_cast<Provider_IExecutionProvider_Router_Impl*>(provider.release()->p_));
+  }
+
+  std::shared_ptr<Provider_IExecutionProviderFactory> p_;
+};
+
+std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_Dnnl(int device_id) {
+#ifdef _WIN32
+  static ProviderLibrary library("onnxruntime_providers_dnnl.dll");
+#else
+  static ProviderLibrary library("libonnxruntime_providers_dnnl.so");
+#endif
+  if (!library.provider_)
+    return nullptr;
+
+  //return std::make_shared<onnxruntime::MkldnnProviderFactory>(device_id);
+  //TODO: This is apparently a bug. The constructor parameter is create-arena-flag, not the device-id
+  return std::make_shared<IExecutionProviderFactory_Translator>(library.provider_->CreateExecutionProviderFactory(device_id));
+}
+
+}  // namespace onnxruntime
+
+ORT_API_STATUS_IMPL(OrtSessionOptionsAppendExecutionProvider_Dnnl, _In_ OrtSessionOptions* options, int use_arena) {
```

Comment:
Should DNNL specific functions be in this file? I would've expected them to under dnnl/.

Reply:
For now, yes. Since this is the core code part that loads the DLL. That can't be inside the provider itself.

We can probably refine this to make it more generic but with only one it's simpler to just hardcode it. I'll generalize it on the second provider.

Reply:
Ok, please add a TODO here.

Reply:
Done